File: 0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution
{
public:
    int maxSubArray(vector<int> &nums)
    {
        int max = -100000;
        int sum = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            sum += nums[i];
            if (sum < nums[i])
            {
                sum = nums[i];
            }
            max = (max > sum ? max : sum);
        }
        return max;
    }
};

File: Python/0053-maximum-subarray.py
# time complexity: O(n)
# space complexity: O(1)
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        result = float('-inf')
        prefix = 0
        for num in nums:
            prefix += num
            prefix = max(prefix, num)
            result = max(result, prefix)
        return result


nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
print(Solution().maxSubArray(nums))
nums = [1]
print(Solution().maxSubArray(nums))
nums = [5, 4, -1, 7, 8]
print(Solution().maxSubArray(nums))
